src/auth/QueueNode.h

```diff
@@ -10,7 +10,9 @@
 #define SQUID_SRC_AUTH_QUEUENODE_H
 
 #include "auth/forward.h"
+#include "auth/UserRequest.h"
```

Comment:
> Basic rule-of-thumb; no header should need to include both a `foo/Foo.h` and `foo/forward.h`.

I am against adopting the above rule of thumb and should requirement because some foo/X.h do _need_ to include foo/forward.h and because that need _changes_.

> The `foo/Foo.h` should be (possibly indirectly) pulling in `foo/forward.h`.

We could adopt a policy that _requires_ all foo/X.h to include foo/forward.h (where the latter exists), but I doubt such a policy is a good idea because it adds costs without offering any significant advantages (that I can see). If somebody thinks it is a good idea, they should propose it, along with pros/cons analysis. The current code does not comply with the above "should" and the corresponding policy. Until we adopt that policy, we cannot rely on it and should not insist on changes based on it.

Files should obviously include the headers they need to pass build tests. Beyond that, we are talking about include list optimization. IMO, include list minimization should not go beyond the first level (i.e. a file including X.h should not rely on X.h including anything else or know what X.h is including) because X.h include list should be X.h own business. This basic encapsulation principle reduces human development time (the primary optimization target) and does not cost much in terms of extra compile time.

Reply:
I agree with Alex here, on pragmatic grounds.
We should try to avoid doing it, but if we have to because our design 
was not the cleanest to start with, it's best to not try and fix everything.

Reply:
> it's best to not try and fix everything

At the very least, it is best not to try to fix something via a constant trickle of polishing change requests based on known-to-be-disputed ideas. We have been here before. Several times. Until we develop a shared model of how the set of include files should be optimized (with respect to nested includes), please reduce overheads by letting the author use whatever they think is best.

FWIW, I am happy to collaborate on establishing an optimization approach we can all agree on (and then enforce).